### ISSUE TITLE
support for zip pack format

### ISSUE THREAD

**Issue by warpfork** [polydawn/rio#3]
We should be able to support zip format.

This should be pretty easy to do, since go stdlib already has a zip format: https://golang.org/pkg/archive/zip/

It will come out as a separate pack implementation, but *most* of the code is probably zimilar to the tar implementation.  They might even be able to share hashing space (though it's unclear that would have an utility other than an amusing coincidence).

Some things like the gradle ecosystem make releases in zip form, so it'd be nice to be able to scan those in easily.

**Comment by willscott** [comment]
it looks annoyingly like there hasn't been any real standardization of xattrs for the zip archive format.
There' s a patch from 2005 proposing support https://www.redhat.com/archives/fedora-selinux-list/2005-October/msg00116.html
but it misuses the intention of keyed extensions in the format, and as such wasn't accepted into infozip https://sourceforge.net/p/infozip/patches/4/

rio could add a custom extension attribute for storing these if we want them as part of a round trip. I've looked through common implementations and haven't found any that keep xattrs.

**Comment by warpfork** [comment]
I missed the xattrs comment until just now, but yeah, xattrs are a real show.  I'm perfectly happy to merge things that don't support them.  The more I learn about xattrs, the more I think that not supporting them might actually be a feature.  I found them perturbing the first time I encountered them, several years ago, and each passing year the impression gets more refined and yet does not improve.

Some fun stuff the clever and almost-painfully-overburdened-with-knowledge-in-this-area @cyphar pointed out to me recently, for example: there exists a set of xattrs prefixed with the string `trusted.`, which cannot even be *seen* in all cases (it may require privileges); many can't be set without various privileges (`security.*`, among others); and apparently there's even some lovely exotica wherein sometimes xattrs when handled from within usernamespaces will get kuids *rewritten*, making them rather fascinatingly tricky to reason about.  (Errors in description of this all mine; I'm going from memory here, which may fail or be imprecise.)

Whatever's going on with xattrs seems to me to be a bunch of stringly-typed nonsense; concretely, it seems one almost has to have a suite of pattern recognition for various *specific* xattrs in order to have a chance at handling them correctly, and treating them like a `map[string]string` will not result in generally correct (or sometimes even *possible*) behavior at all... so I find myself seriously asking if there's much point in this (and if implementations of tools out there which _pretend_ xattrs can be treated as `map[string]string` might not in turn out to be rather dangerous!).  Furthermore, even if we went in for such a table of special patterns, and got such a table right at one point in time, it would surely  become driftingly wrong across kernel versions; this seems to be an area where exciting new "features" get shoved with far less review than would be the standard for other parts of the kernel surface area.

No... long story short, I'm quite alright with merging things that simply do not regard xattrs.

This might be worth a fuller write-up somewhere (ideally version controlled), but I thought since it was mentioned here and we just flipped this issue to done, it's worth a quick jot here and now.
